### ISSUE TITLE
Can't read out values after disconnect

### ISSUE THREAD

**Issue by everlof** [NordicSemiconductor/IOS-CoreBluetooth-Mock#98]
Hello,

I've integrated this framework and it works pretty well. However, when we reconnect, we don't get callbacks for the for the `didUpdateValueFor` after we've called `readValue(for characteristic`.

If I just restart the app, it works fine and connect. Do you have any hint of what to look for and how to debug this issue?

To clarify. This only happens while running without mock enabled. Mocking works good and running without this library work good as well.

What we do is:

1. Call connect after the device reports that it disconnected
2. Get "didConnect" callback
3. Call "readValue" on the peripheral.

**Comment by philips77** [comment]
Hi, thank you for investigation.
Do you have the issue when using mock or native variant of `CBMCentralManager`?

**Comment by everlof** [comment]
I've got the issue when using `CBMCentralManagerNative`

**Comment by everlof** [comment]
With the risk of repeating myself, it seems like it's not 100% correct to cache the services? I mean, since if I run purely on the `CBCentralManager`, the `CBPeripheral` received in `didConnect` doesn't have any services when it reconnects.

**Comment by everlof** [comment]
Perhaps this is an even more sensible solution?

```
diff --git a/CoreBluetoothMock/CBMCentralManagerNative.swift b/CoreBluetoothMock/CBMCentralManagerNative.swift
index 761a95d..8439bdf 100644
--- a/CoreBluetoothMock/CBMCentralManagerNative.swift
+++ b/CoreBluetoothMock/CBMCentralManagerNative.swift
@@ -89,6 +89,7 @@ public class CBMCentralManagerNative: CBMCentralManager {
             manager.delegate?.centralManager(manager,
                                              didDisconnectPeripheral: getPeripheral(peripheral),
                                              error: error)
+            removePeripheral(peripheral)
         }
         
         #if !os(macOS)
@@ -120,6 +121,10 @@ public class CBMCentralManagerNative: CBMCentralManager {
             manager.peripherals[peripheral.identifier] = p
             return p
         }
+
+        private func removePeripheral(_ peripheral: CBPeripheral) {
+            manager.peripherals[peripheral.identifier] = nil
+        }
     }
     
     private class CBMCentralManagerDelegateWrapperWithRestoration: CBMCentralManagerDelegateWrapper {
```

**Comment by philips77** [comment]
Thanks for the diff. I hope I'll find some time for this issue before my vacations. If not, it will never to wait few weeks.

**Comment by everlof** [comment]
Any updates regarding this?

**Comment by philips77** [comment]
Hi, sorry. I didn't have time to look into this library yet. @NickKibish has released version 0.17.0 but with other changes.
I hope me or Nick will find some time this month.

**Comment by everlof** [comment]
Hello! Just wondering if anything's done regarding this? Or it's just postponed?

**Comment by philips77** [comment]
Hello, we're busy with other tasks. This library is waiting for its turn. Sorry for the delay. Could you propose a change in a PR with a fix, that would be way faster.

**Comment by everlof** [comment]
Absolutely, I just asked since you hinted that it might be within a month.

I've created an MR: https://github.com/NordicSemiconductor/IOS-CoreBluetooth-Mock/pull/102

**Comment by philips77** [comment]
Merged and released as 0.18.0.
